File: BlockGridKids/Model/Board.swift
import Foundation

/// The square playfield. Each cell either is empty (`nil`) or holds a color index.
struct Board: Codable, Equatable {

    /// Board sizes the player can pick between in settings.
    static let availableSizes = [5, 8]
    /// The size used for a fresh install and for saves written before the
    /// board-size setting existed.
    static let defaultSize = 8

    /// Number of rows and columns. The board is always square.
    let size: Int

    /// Row-major storage of `size * size` cells.
    private(set) var cells: [Int?]

    init(size: Int = Board.defaultSize) {
        self.size = Board.availableSizes.contains(size) ? size : Board.defaultSize
        cells = Array(repeating: nil, count: self.size * self.size)
    }

    // MARK: - Codable

    private enum CodingKeys: String, CodingKey {
        case size
        case cells
    }

    /// Decoding tolerates a missing `size` so a game saved by an older build
    /// still restores as an 8x8 board instead of being thrown away.
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let decodedSize = try container.decodeIfPresent(Int.self, forKey: .size) ?? Board.defaultSize
        let decodedCells = try container.decode([Int?].self, forKey: .cells)
        guard Board.availableSizes.contains(decodedSize),
              decodedCells.count == decodedSize * decodedSize else {
            throw DecodingError.dataCorruptedError(
                forKey: .cells,
                in: container,
                debugDescription: "Cell count does not match the board size"
            )
        }
        size = decodedSize
        cells = decodedCells
    }

    // MARK: - Cell access

    func isInBounds(_ position: GridPosition) -> Bool {
        position.row >= 0 && position.row < size && position.col >= 0 && position.col < size
    }

    private func index(of position: GridPosition) -> Int {
        position.row * size + position.col
    }

    subscript(position: GridPosition) -> Int? {
        get {
            guard isInBounds(position) else { return nil }
            return cells[index(of: position)]
        }
        set {
            guard isInBounds(position) else { return }
            cells[index(of: position)] = newValue
        }
    }

    func isEmpty(at position: GridPosition) -> Bool {
        guard isInBounds(position) else { return false }
        return cells[index(of: position)] == nil
    }

    var isCompletelyEmpty: Bool {
        cells.allSatisfy { $0 == nil }
    }

    var filledCellCount: Int {
        cells.reduce(into: 0) { total, cell in
            if cell != nil { total += 1 }
        }
    }

    // MARK: - Placement

    /// The board positions a piece would occupy if anchored at `origin`.
    /// `origin` maps to the shape's normalized `(row: 0, col: 0)` corner.
    func positions(for shape: ShapeTemplate, at origin: GridPosition) -> [GridPosition] {
        shape.offsets.map { origin.offset(byRow: $0.row, col: $0.col) }
    }

    /// True when every cell the piece would occupy is in bounds and empty.
    func canPlace(_ shape: ShapeTemplate, at origin: GridPosition) -> Bool {
        for position in positions(for: shape, at: origin) {
            guard isInBounds(position) else { return false }
            guard cells[index(of: position)] == nil else { return false }
        }
        return true
    }

    /// True when the shape fits anywhere at all on the current board.
    func canPlaceAnywhere(_ shape: ShapeTemplate) -> Bool {
        firstValidOrigin(for: shape) != nil
    }

    /// The topmost-leftmost origin where the shape fits, if any.
    func firstValidOrigin(for shape: ShapeTemplate) -> GridPosition? {
        let maxRow = size - shape.height
        let maxCol = size - shape.width
        guard maxRow >= 0, maxCol >= 0 else { return nil }
        for row in 0...maxRow {
            for col in 0...maxCol {
                let origin = GridPosition(row: row, col: col)
                if canPlace(shape, at: origin) {
                    return origin
                }
            }
        }
        return nil
    }

    /// Fills the piece's cells with `colorIndex`. Returns the filled positions.
    @discardableResult
    mutating func place(_ shape: ShapeTemplate, at origin: GridPosition, colorIndex: Int) -> [GridPosition] {
        let filled = positions(for: shape, at: origin)
        for position in filled {
            cells[index(of: position)] = colorIndex
        }
        return filled
    }

    // MARK: - Clearing

    /// Rows and columns that are currently completely filled.
    func completedLines() -> (rows: [Int], columns: [Int]) {
        var rows: [Int] = []
        var columns: [Int] = []

        for row in 0..<size {
            let isFull = (0..<size).allSatisfy { cells[row * size + $0] != nil }
            if isFull { rows.append(row) }
        }
        for col in 0..<size {
            let isFull = (0..<size).allSatisfy { cells[$0 * size + col] != nil }
            if isFull { columns.append(col) }
        }
        return (rows, columns)
    }

    /// Rows and columns that *would* be completed by dropping `shape` at
    /// `origin`, without mutating anything. Drives the drag-time hint that
    /// shows the player which lines a move is about to clear.
    func linesCompletedIfPlaced(_ shape: ShapeTemplate, at origin: GridPosition) -> (rows: [Int], columns: [Int]) {
        guard canPlace(shape, at: origin) else { return ([], []) }
        let added = Set(positions(for: shape, at: origin))

        func isFilled(row: Int, col: Int) -> Bool {
            cells[row * size + col] != nil || added.contains(GridPosition(row: row, col: col))
        }

        var rows: [Int] = []
        var columns: [Int] = []
        for row in 0..<size where (0..<size).allSatisfy({ isFilled(row: row, col: $0) }) {
            rows.append(row)
        }
        for col in 0..<size where (0..<size).allSatisfy({ isFilled(row: $0, col: col) }) {
            columns.append(col)
        }
        return (rows, columns)
    }

    /// Empties the given rows and columns. Returns every cleared position,
    /// de-duplicated so intersections are only reported once.
    @discardableResult
    mutating func clear(rows: [Int], columns: [Int]) -> [GridPosition] {
        var cleared: Set<GridPosition> = []

        for row in rows {
            for col in 0..<size {
                cleared.insert(GridPosition(row: row, col: col))
            }
        }
        for col in columns {
            for row in 0..<size {
                cleared.insert(GridPosition(row: row, col: col))
            }
        }
        for position in cleared {
            cells[index(of: position)] = nil
        }
        return Array(cleared)
    }

    mutating func removeAll() {
        cells = Array(repeating: nil, count: size * size)
    }
}

File: web/js/board.js
// The square playfield. Ported from BlockGridKids/Model/Board.swift.
import { DEFAULT_BOARD_SIZE } from './shapes.js';

const AVAILABLE_SIZES = [5, 8];

function key(pos) {
  return `${pos.row},${pos.col}`;
}

class Board {
  constructor(size = DEFAULT_BOARD_SIZE) {
    this.size = AVAILABLE_SIZES.includes(size) ? size : DEFAULT_BOARD_SIZE;
    // Row-major storage; each cell is null (empty) or a color index.
    this.cells = new Array(this.size * this.size).fill(null);
  }

  static get availableSizes() {
    return AVAILABLE_SIZES;
  }
  static get defaultSize() {
    return DEFAULT_BOARD_SIZE;
  }

  isInBounds(pos) {
    return pos.row >= 0 && pos.row < this.size && pos.col >= 0 && pos.col < this.size;
  }

  index(pos) {
    return pos.row * this.size + pos.col;
  }

  get(pos) {
    if (!this.isInBounds(pos)) return null;
    return this.cells[this.index(pos)];
  }

  set(pos, value) {
    if (!this.isInBounds(pos)) return;
    this.cells[this.index(pos)] = value;
  }

  isEmpty(pos) {
    if (!this.isInBounds(pos)) return false;
    return this.cells[this.index(pos)] === null;
  }

  get isCompletelyEmpty() {
    return this.cells.every((c) => c === null);
  }

  get filledCellCount() {
    return this.cells.reduce((t, c) => (c !== null ? t + 1 : t), 0);
  }

  // The positions a shape would occupy if anchored at origin.
  positions(shape, origin) {
    return shape.offsets.map((o) => ({ row: origin.row + o.row, col: origin.col + o.col }));
  }

  canPlace(shape, origin) {
    for (const pos of this.positions(shape, origin)) {
      if (!this.isInBounds(pos)) return false;
      if (this.cells[this.index(pos)] !== null) return false;
    }
    return true;
  }

  canPlaceAnywhere(shape) {
    return this.firstValidOrigin(shape) !== null;
  }

  firstValidOrigin(shape) {
    const maxRow = this.size - shape.height;
    const maxCol = this.size - shape.width;
    if (maxRow < 0 || maxCol < 0) return null;
    for (let row = 0; row <= maxRow; row++) {
      for (let col = 0; col <= maxCol; col++) {
        const origin = { row, col };
        if (this.canPlace(shape, origin)) return origin;
      }
    }
    return null;
  }

  place(shape, origin, colorIndex) {
    const filled = this.positions(shape, origin);
    for (const pos of filled) this.cells[this.index(pos)] = colorIndex;
    return filled;
  }

  completedLines() {
    const rows = [];
    const columns = [];
    for (let row = 0; row < this.size; row++) {
      let full = true;
      for (let c = 0; c < this.size; c++) {
        if (this.cells[row * this.size + c] === null) { full = false; break; }
      }
      if (full) rows.push(row);
    }
    for (let col = 0; col < this.size; col++) {
      let full = true;
      for (let r = 0; r < this.size; r++) {
        if (this.cells[r * this.size + col] === null) { full = false; break; }
      }
      if (full) columns.push(col);
    }
    return { rows, columns };
  }

  // Lines that would complete if the shape were placed, without mutating.
  linesCompletedIfPlaced(shape, origin) {
    if (!this.canPlace(shape, origin)) return { rows: [], columns: [] };
    const added = new Set(this.positions(shape, origin).map(key));
    const isFilled = (row, col) =>
      this.cells[row * this.size + col] !== null || added.has(`${row},${col}`);

    const rows = [];
    const columns = [];
    for (let row = 0; row < this.size; row++) {
      let full = true;
      for (let c = 0; c < this.size; c++) if (!isFilled(row, c)) { full = false; break; }
      if (full) rows.push(row);
    }
    for (let col = 0; col < this.size; col++) {
      let full = true;
      for (let r = 0; r < this.size; r++) if (!isFilled(r, col)) { full = false; break; }
      if (full) columns.push(col);
    }
    return { rows, columns };
  }

  // Empties rows/columns; returns every cleared position, de-duplicated.
  clear(rows, columns) {
    const cleared = new Map();
    for (const row of rows) {
      for (let col = 0; col < this.size; col++) cleared.set(`${row},${col}`, { row, col });
    }
    for (const col of columns) {
      for (let row = 0; row < this.size; row++) cleared.set(`${row},${col}`, { row, col });
    }
    for (const pos of cleared.values()) this.cells[this.index(pos)] = null;
    return Array.from(cleared.values());
  }

  removeAll() {
    this.cells = new Array(this.size * this.size).fill(null);
  }

  // Serialization helpers for save/restore.
  toJSON() {
    return { size: this.size, cells: this.cells };
  }

  static fromJSON(obj) {
    if (!obj) return null;
    const size = AVAILABLE_SIZES.includes(obj.size) ? obj.size : DEFAULT_BOARD_SIZE;
    if (!Array.isArray(obj.cells) || obj.cells.length !== size * size) return null;
    const board = new Board(size);
    board.cells = obj.cells.map((c) => (c === null || c === undefined ? null : c));
    return board;
  }
}

export { Board };
